_drafts/practical_zephyr_01_kconfig.md

```diff
@@ -0,0 +1,986 @@
+
+```conf
+CONFIG_PRINTK=n
+```
+
+What's the effect of changing a symbol? As mentioned before, this _can_ affect the build options and _can_ also be used by some subsystem's implementation. Whatever a symbol does, however, is specific for each symbol, meaning there is no convention for symbol names that always have the same effect. Thus, in the end, it is up to you to dig into the sources and documentation to see which options are available and what effect they have (there's a [Kconfig search](#kconfig-search) that we'll see later).
+
+Let's have a short look at our `PRINTK` symbol. A quick search for `CONFIG_PRINTK` in our local Zephyr code reveals how the symbol is used in the implementation of the `printk` debugging subsystem. Notice that the below snippets are simplified versions of the real code:
+
+`zephyr/lib/os/printk.h`
+```c
+#ifdef CONFIG_PRINTK
+extern void printk(const char *fmt, ...);
+#else
+static inline void printk(const char *fmt, ...)
+{
+	ARG_UNUSED(fmt);
+}
+```
+
+`zephyr/lib/os/printk.c`
+```c
+#ifdef CONFIG_PRINTK
+void printk(const char *fmt, ...)
+{
+  // ...
+}
+#endif /* defined(CONFIG_PRINTK) */
+```
+
+> **Notice:** In our `prj.conf` file we **set** `CONFIG_PRINTK=n`, but in the implementation, this translates to a missing definition, not a definition with a value of `n`. This is simply what a `bool` symbol translates to in the implementation. For details, I'll once again have to direct you to the [official documentation](https://docs.zephyrproject.org/latest/build/kconfig/setting.html#setting-kconfig-configuration-values).
+
+In the header file `zephyr/lib/os/printk.h` we can see that `printk` is replaced by a dummy function in case `PRINTK` is disabled. Thus, all the calls to `printk` will effectively be removed by the compiler. In the `zephyr/lib/os/printk.c` source file we can see that the `#if` directive is used to conditionally compile the relevant source code only in case `PRINTK` is enabled. Later in this section, we'll also see an example of how files can be excluded entirely by the build system in case an option is not enabled.
+
+With `printk` disabled in our application configuration file, let's try to rerun our application to verify that the output is indeed disabled:
+
+```bash
+$ west build --board nrf52840dk_nrf52840 --build-dir ../build --pristine
+$ west flash --build-dir ../build
+```
+
+> **Notice:** When changing your _Kconfig_ symbols, it is recommended to use the `--pristine` option to enforce a complete rebuild of your application. This in turn also requires specifying the correct `--board` in case you didn't configure it using `west config`.
```

Comment:
You're definitely right! I mostly fall back to `--pristine` since I'm not yet very sure when it's required. Happy that you're pointing out the places where it isn't necessary, thanks a lot!